### PR TITLE
revise room.say() mention function

### PR DESCRIPTION
**Pull request by windmemory**
## I'm submitting a...

```
[ ] Bug Fix
[ ] Feature
[x] Other (Refactoring, Added tests, Documentation, ...)
```
Refer to https://github.com/Chatie/wechaty/issues/1718

This PR involves a **BREAKING CHANGE**

### CODE REVIEW

**Comment by windmemory** [comment]
## BREAKING CHANGE:

`room.say()` function no longer accept arguments like below:
```ts
room.say(text: string, mention: Contact[])
```
## Alternative solution
To mention multiple contacts when sending out text messages, there are two new ways of doing that:
#### `room.say(text: string, ...mentionList: Contact[])`
Sample code:
```ts
// Assume you got there Contact instance
// contactA assume the name is Alex
// contactB assume the name is Bob
// contactC assume the name is Carson
await room.say('let's party', contactA, contactB, contactC)
```
Send out text will be 
```
@Alex @Bob @Carson let's party
```

#### Tagged Template
Sample code:
```ts
// Assume you got there Contact instance
// contactA assume the name is Alex
// contactB assume the name is Bob
// contactC assume the name is Carson
room.say`Hello ${contactA} and ${contactB}, this is ${contactC}`
```
Send out text will be 
```
Hello @Alex and @Bob, this is @Carson
```

File: src/user/room.spec.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@
 import test  from 'blue-tape'
 import sinon from 'sinon'
 
-import { RoomPayload }  from 'wechaty-puppet'
+import { ContactPayload, RoomMemberPayload, RoomPayload }  from 'wechaty-puppet'
```
Comment (huan) [inline src/user/room.spec.ts]:
```
import { 
  ContactPayload,
  RoomMemberPayload,
  RoomPayload,
}                            from 'wechaty-puppet'
```

File: src/user/room.spec.ts [hunk 2]
```diff
@@ -54,3 +54,75 @@ test('findAll()', async t => {
...
+  const contact2 = wechaty.Contact.load(EXPECTED_CONTACT_2_ID)
+  await contact1.sync()
+  await contact2.sync()
+  await room.sync()
+  await room.say`To be ${contact1} or not to be ${contact2}`
+  await room.say('Yo', contact1)
+  await room.say('hey buddies, let\'s party', contact1, contact2)
+  await room.say('')
+
+  t.deepEqual(callback.getCall(0).args, [
```
Comment (huan) [inline src/user/room.spec.ts]: Please add subtests for each `room.say` test.

i.e.

```ts
test('say()', async t => {
  // tear up...
  test('say() with param1', async t => {
    // do say() test 1
  })
  test('say() with param2', async t => {
    // do say() test 2
  })
  // other subtests ...
})
```

File: src/user/room.ts [hunk 3]
```diff
@@ -425,63 +425,106 @@ export class Room extends Accessory implements Sayable {
...
+      }, textOrListOrContactOrFileOrUrl.id)
+    } else if (textOrListOrContactOrFileOrUrl instanceof UrlLink) {
       /**
        * 4. Link Message
        */
       await this.puppet.messageSendUrl({
         contactId : this.id
-      }, textOrContactOrFileOrUrl.payload)
+      }, textOrListOrContactOrFileOrUrl.payload)
+    } else if (textOrListOrContactOrFileOrUrl instanceof Array) {
```
Comment (huan) [inline src/user/room.ts]: Please add a sub-function for the following codes for `sayTemplateStringsArray()`.

```ts
    } else if (textOrListOrContactOrFileOrUrl instanceof Array) {
      return sayTemplateStringsArray(
        textOrListOrContactOrFileOrUrl, 
        ...mentionList,
      )
    }
```

**Comment by windmemory** [comment]
@huan Done.

**Comment by huan** [comment]
Thanks for your great PR!

**Pull request merged**